azure/services/availabilitysets/spec.go

```diff
@@ -0,0 +1,85 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package availabilitysets
+
+import (
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2021-04-01/compute"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/pkg/errors"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-azure/azure/converters"
+)
+
+// AvailabilitySetSpec defines the specification for an availability set.
+type AvailabilitySetSpec struct {
+	Name             string
+	ResourceGroup    string
+	ClusterName      string
+	Location         string
+	FaultDomainCount *int32
+	AdditionalTags   infrav1.Tags
+}
+
+// ResourceName returns the name of the availability set.
+func (s *AvailabilitySetSpec) ResourceName() string {
+	return s.Name
+}
+
+// ResourceGroupName returns the name of the resource group.
+func (s *AvailabilitySetSpec) ResourceGroupName() string {
+	return s.ResourceGroup
+}
+
+// OwnerResourceName is a no-op for availability sets.
+func (s *AvailabilitySetSpec) OwnerResourceName() string {
+	return ""
+}
+
+// Parameters returns the parameters for the availability set.
+func (s *AvailabilitySetSpec) Parameters(existing interface{}) (interface{}, error) {
+	if existing != nil {
+		if _, ok := existing.(compute.AvailabilitySet); !ok {
+			return nil, errors.Errorf("%T is not a compute.AvailabilitySet", existing)
+		}
+		// availability set already exists
+		return nil, nil
+	}
+
+	if s.FaultDomainCount == nil {
+		return nil, errors.New("unable to find required FaultDomainCount from machine cache")
+	}
+
+	asParams := compute.AvailabilitySet{
+		// TODO: should availability sets be given a name?
```

Comment:
the name is part of the CreateOrUpdate function parameters (same with the resource group)